src/main/java/org/commcare/formplayer/beans/menus/EntityListResponse.java

```diff
@@ -96,22 +96,19 @@ public EntityListResponse(EntityScreen nextScreen,
             entities = processEntitiesForCaseDetail(detail, reference, ec, neededDatum);
         } else {
             Vector<TreeReference> references = nextScreen.getReferences();
-            List<EntityBean> entityList = processEntitiesForCaseList(detail, references, ec,
-                    searchText, neededDatum, sortIndex, isFuzzySearchEnabled);
+            List<Entity<TreeReference>> entityList = buildEntityList(detail, ec, references, searchText,
+                    sortIndex, isFuzzySearchEnabled);
 
             if (casesPerPage == 0) {
                 casesPerPage = DEFAULT_CASES_PER_PAGE;
             }
             casesPerPage = Math.min(casesPerPage, MAX_CASES_PER_PAGE);
 
-            if (entityList.size() > casesPerPage && !(detail.getNumEntitiesToDisplayPerRow() > 1)) {
-                // we're doing pagination
-                setCurrentPage(offset / casesPerPage);
-                setPageCount((int)Math.ceil((double)entityList.size() / casesPerPage));
-                entityList = paginateEntities(entityList, casesPerPage, offset);
-            }
-            entities = new EntityBean[entityList.size()];
-            entityList.toArray(entities);
+            List<Entity<TreeReference>> entitesForPage = paginateEntities(entityList, detail, casesPerPage,
+                    offset);
+            List<EntityBean> entityBeans = processEntitiesForCaseList(detail, entitesForPage, ec, neededDatum);
+            entities = new EntityBean[entityBeans.size()];
+            entityBeans.toArray(entities);
```

Comment:
@knollpointer the 'processing' is split into two stages:
1. get the list of entities
2. convert each entity to the format required for display

Previously the pagination was happening after step 2 but this change makes it happen in between stage 1 & 2 which means only the paginated list is converted to display objects.

Reply:
> if buildEntityList is running all the calcs on the entire result set anyways, 

@knollpointer That's a great observation. It seems like earlier we were calculating the detail fields twice as part of 'processing'  for the entire case list and now after this PR we are not doing it twice for all cases but only for the cases on the current page which is still bad. I have opened up [another PR](https://github.com/dimagi/formplayer/pull/1256) to avoid this duplicate calculation even for the cases on the current page. Thanks for raising this and talking through this with me!

Reply:
@knollpointer The new PR is on staging already in case you want to test it and see if there is any improvement in performance. I would only hope for a slight increase in improvement due to this change.

Reply:
👍🏻 I didn't realize that `nodeEntityFactory.getEntity` was actually building the full dataset.